### ISSUE TITLE
HTTP2 Support

### ISSUE THREAD

**Issue by JesseFarebro** [emilevauge/traefik#139]
I was just wondering whether HTTP2 support was on the roadmap anywhere? I ask this as I would love to load balance gRPC with Traefik.

**Comment by txgruppi** [comment]
Go 1.6's `net/http` will be HTTP2 right? Maybe we will get HTTP2 support for free.

**Comment by emilevauge** [comment]
It's even possible to use `golang.org/x/net/http2` in go 1.5 ;)

**Comment by strarsis** [comment]
There is a golang nightly image on dockerhub, you can use it to build the source from that directory:
````
$ mkdir traefik-build && cd traefi-build
$ docker run --rm -v $(pwd):/root/src strarsis/golang-build-1.5.2-nightly-a766d8e59c2c43b4c6795404d303b64b9fe69a0f /bin/sh -c \
'cd /root/src && export GOPATH=$(pwd) && go get -u github.com/emilevauge/traefik'
````
From previous experience, HTTP2 should work out of the box when built with go 1.6 - 
you have to use TLS/SSL though because most if not all HTTP2 clients require this (e.g. nghttp).

With best regards

**Comment by f3l1x** [comment]
Could you please help me to setup up http2? Is there any special configuration for that? This tester (https://tools.keycdn.com/http2-test) told me it's not http2. Thank you for helping me.

**Comment by tlvenn** [comment]
What is actually needed to enable HTTP2 beyond TLS on Traefik side ? Is there any requirement on backends ? I was expecting this to work outside of the box as soon as content was served over https but as @f3l1x outlined, this is not the case.

**Comment by trecloux** [comment]
@tlvenn You do not need to configure anything more than activating TLS.
Version 1.1.0-rc3 had an issue with HTTP/2, this was fixed with 1.1.0-rc4.
You can see a demo there : https://github.com/trecloux/traefik-http2-demo

**Comment by tlvenn** [comment]
Ha thanks @trecloux ! Was running on rc3, updated to rc4 and now HTTP2 is enabled 🎉

**Comment by andrewwebber** [comment]
@trecloux @tlvenn 
Has anyone tested this with gRPC?

I cannot proxy a gRPC backend and get the following error:
gRPC requires HTTP/2
```bash
Framer 0xc4200740d0: read DATA flags=END_STREAM stream=1 len=21 data="gRPC requires HTTP/2\n"
```

This is because the backend is called with http version 1

When comparing a gRPC client connecting directly to the backend there is the following log difference backend side

gRPC client -> gRPC backend
content-type header: application/grpc, method: POST, **proto-major: 2**, content-type: application/grpc

gRPC client -> traefik -> gRPC backend
content-type header: application/grpc, method: POST, **proto-major: 1**, content-type: application/grpc

The following pattern is commonly used to have an http endpoint that supports http V1 and gRPC
```golang
func grpcHandlerFunc(rpcServer *grpc.Server, otherHandler http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		log.Printf("content-type header: %v, method: %v, proto-major: %v, content-type: %s", r.Header.Get("Content-Type"), r.Method, r.ProtoMajor, r.Header.Get("Content-Type"))

		if r.ProtoMajor == 2 && (strings.Contains(r.Header.Get("Content-Type"), "application/grpc") || r.Method == "PRI") {
			log.Printf("handling gRPC request")
			rpcServer.ServeHTTP(w, r)
			return
		}

		log.Printf("handling regular HTTP1.x/2 request")
		otherHandler.ServeHTTP(w, r)
	})
}
```

This logic works with nghttpx proxy

**Comment by jhump** [comment]
@andrewwebber, what version of Go were you using? According to golang/go#14391 (which I suspect is the underlying problem), this should be fixed in 1.6.2 and 1.7.

I'm wondering because I also want to use traefik to proxy grpc traffic. I haven't tried it yet: I figured I'd search for known issues before spending time experimenting.

**Comment by ldez** [comment]
@jhump you can follow this issue #1419
